### ISSUE TITLE
"overload signature is not compatible" shows up more than once in the LS

### ISSUE THREAD

**Issue by jakebailey** [microsoft/TypeScript#49916]
# Bug Report

### 🔎 Search Terms

overload signature is not compatible language server vs code vscode diagnostic problem error

### 🕗 Version & Regression Information

Probably #40094. But, old TS had one error in 3.3, then 2 in 3.5, then 7 in 4.1+, so this isn't just one PR.

### ⏯ Playground Link

```ts repro
// @strict: true
// @errors: 2394

export interface Push<T> {
    push(...values: T[]): void;
    /* @internal*/ readonly length: number;
}

export function append<TArray extends any[] | undefined, TValue extends NonNullable<TArray>[number] | undefined>(to: TArray, value: TValue): [undefined, undefined] extends [TArray, TValue] ? TArray : NonNullable<TArray>[number][];
export function append<T>(to: T[], value: T | undefined): T[];
export function append<T>(to: T[] | undefined, value: T): T[];
export function append<T>(to: T[] | undefined, value: T | undefined): T[] | undefined;
export function append<T>(to: Push<T>, value: T | undefined): void;
export function append<T>(to: T[], value: T | undefined): T[] | undefined {
    if (value === undefined) return to;
    if (to === undefined) return [value];
    to.push(value);
    return to;
}

```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?ts=4.8.0-dev.20220714#code/PTAEAEGcBcCcEsDG0Bco4FcCmAoEEtZYB7WSNAJgGYBOAFhxywA8AHU6UeAO2kIDMAhoiygAChkgALADwAVAHygA3jlDrQrSVIAUAOgMA3QQBts5UHIDaAXQCUaQ8XgATANxqNwAFQQefWG5TbzBYLEEXYm4TAE9QEyxuAHNoKTRuDABbACNCDwBfRhZ2WE5+DG5keCjQQVZWRJd5AEEiQTiWPm4XSFruGNtQAB9QCpcsfh4sFwAaSwA1U2xQTsbegDko9YwTE0FshJa2mIUrDJzCG2HR7omplwUdaGI0OVbYdrnjMyxXxZ+HKArGM7txpnMQZMwS4rqtur0rG9jnM5P9sFcAPyWd7tUBoTbcba7faHJEfE5nLK5WA2WweYocUDlSrQarcWr1RryR7PV62L5LX6Wa6Q+6A6w2elsRnMqo1OoNbrcp4vSyDEai6ECn6vcV0pjS0pMipy9kKrmKFV8q4a25Q8Ggb7YV4iu1i62u8b29wGkplE2s+WcpWW3nibTc7XO4W2r3ux3OH0Mo2ywNm4NNUOqiVRoVyT2g6Z6m03OPQlSedTwfigHRO0QAXibpcLLjsoDC0AwgXQxA8Gi4NZVoCbDZb3vbne77Ks9clld7ei00jrgrs-Y0U57zwKOCAA)

[Playground Link](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgBQFcBnACwB4AVAPjgG8AoOZuME0gCgDoeA3TADaFgxAFxxKAbQC6ASnG8ICACYBuJiwD0AKjgABZKihJB2zXCjBMyiEgEBPOAOBIA5jFLikhALYAjNHUAXwYGUEhYOHRCJGwYBFs4TDAwF2UqAEEoKExHUFQkZWIkpHsZOAAfOBjlYHRkYGUAGgkANUFhOHy04oA5W17CAQFMP2dM7NzqSW9-NGlK6sK6huVqDhgIcUosnPsW-iFgbfaj+ThJGpWkRpar+pvlBe7C4skdyf22juAFgH4JLtcnBxP0kINhqNxh89tNZgEoNIZOpwtB4NFYvFEslUoUqOtNtsZAcfttFvdVucpNIUeA0VEYnEEkgkik0viNlsJOUqhTHiSjtsqciwnTIhimdi2XiaJyiQtecsHrc4IdhGTFbVlcphQqllrVrSIujGViWTj2bLCQR2PiBeqJOSlZSFEo1KLjQzMczWbj0laudT7cdHZrro1dU6DY96BpmAh0HAOGq8ABedP68M6izAGCEYxwTbqFiIROcuDp1OZ7WyHN5guSFM0uOFiBcNhkZM-WTFliWessosMEJAA)

### 🙁 Actual behavior

Many, many errors, but only in the LS.

![image](https://user-images.githubusercontent.com/5341706/179132223-75062797-b230-4730-8f32-c1a2a29b163d.png)

### 🙂 Expected behavior

One error about overload signatures not being correct.

**Comment by jakebailey** [comment]
The errors go away if you get rid of `export` on everything, and #40094 made the check always occur: https://github.com/microsoft/TypeScript/pull/40094/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R32219

Which is probably the problem, though I am confused as to how this error is not suppressed for being duplicated. Or why all positions are the same when it's each individual overload that should be checked.

**Comment by jakebailey** [comment]
Thankfully testable using fourslash (but not a compiler test):


```
Semantic Diagnostics for file '/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts':
/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts(8,17): error TS2394: This overload signature is not compatible with its implementation signature.
/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts(8,17): error TS2394: This overload signature is not compatible with its implementation signature.
/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts(8,17): error TS2394: This overload signature is not compatible with its implementation signature.
/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts(8,17): error TS2394: This overload signature is not compatible with its implementation signature.
/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts(8,17): error TS2394: This overload signature is not compatible with its implementation signature.
/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts(8,17): error TS2394: This overload signature is not compatible with its implementation signature.
/tests/cases/fourslash/overloadNotCompatibleErrorDuplication.ts(8,17): error TS2394: This overload signature is not compatible with its implementation signature.
```

**Comment by jakebailey** [comment]
@typescript-bot bisect good v3.3.3 bad v3.5.1

**Comment by jakebailey** [comment]
Turns out this was fixed by #50309 (which I think has enough test coverage).
